### PR TITLE
Encapsulate webhook controller arguments into a single class

### PR DESCRIPTION
**Pull request by alex-ko-dev**
### Description

Returning technical debt: webhook handler methods had lots of arguments. These arguments are bound to request query string. This made the controller code difficult to maintain and read. 

The main changes:
- extract arguments that are common for the controllers into a separate class (ControllerRequest)
- extract the common arguments of ADO/TFS controllers into the AdoDetailsRequest class
- extract duplicate code into base classes: WebhookController and AdoControllerBase

Some query string parameters are defined using kebab case, e.g. `exclude-files`. Spring Boot doesn't support binding such parameters to class fields out of the box. To fix this, CaseTransformingFilter was added.

### References
[Work item 154](https://dev.azure.com/CxFlow/CxFlow/_workitems/edit/154/).

### Testing

The changes are covered by existing tests that involve calls to webhook controllers.

### CODE REVIEW

**Comment by cxflowtestuser** [comment]
Scan submitted to Checkmarx

**Comment by cxflowtestuser** [comment]
### Checkmarx scan completed
[Full Scan Details](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010448&projectid=154)
#### Checkmarx Scan Summary
Severity|Count
---|---
High|0
Medium|112
Low|1408
Informational|56

#### Violation Summary
Severity|Count
---|---

#### Details
|Lines|Severity|Category|File|Link|
---|---|---|---|---

File: src/main/java/com/checkmarx/flow/CxFlowRunner.java [hunk 1]
```diff
@@ -1,10 +1,7 @@
 package com.checkmarx.flow;
 
 import com.checkmarx.flow.config.*;
-import com.checkmarx.flow.dto.BugTracker;
-import com.checkmarx.flow.dto.ExitCode;
-import com.checkmarx.flow.dto.FlowOverride;
-import com.checkmarx.flow.dto.ScanRequest;
+import com.checkmarx.flow.dto.*;
```
Comment (ofersk) [inline src/main/java/com/checkmarx/flow/CxFlowRunner.java:4]: please avoid using * in imports

File: src/main/java/com/checkmarx/flow/config/FlowConfig.java [hunk 2]
```diff
@@ -11,7 +12,7 @@
 import org.springframework.web.client.RestTemplate;
 
 import java.beans.ConstructorProperties;
-import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
```
Comment (ofersk) [inline src/main/java/com/checkmarx/flow/config/FlowConfig.java:15]: is this deliberate?

File: src/main/java/com/checkmarx/flow/controller/ADOController.java [hunk 3]
```diff
@@ -3,40 +3,35 @@
 import com.checkmarx.flow.config.ADOProperties;
 import com.checkmarx.flow.config.FlowProperties;
 import com.checkmarx.flow.config.JiraProperties;
-import com.checkmarx.flow.dto.BugTracker;
-import com.checkmarx.flow.dto.EventResponse;
-import com.checkmarx.flow.dto.FlowOverride;
-import com.checkmarx.flow.dto.ScanRequest;
+import com.checkmarx.flow.dto.*;
```
Comment (ofersk) [inline src/main/java/com/checkmarx/flow/controller/ADOController.java:6]: avoid * in imports

File: src/main/java/com/checkmarx/flow/controller/ADOController.java [hunk 4]
```diff
@@ -50,86 +45,60 @@
...
+            if (repository.getName().startsWith(properties.getTestRepository())) {
                 log.info("Handling ADO Test Event");
                 return ResponseEntity.status(HttpStatus.OK).body(EventResponse.builder()
                         .message("Test Event").success(true).build());
             }
 
-            if(!ScanUtils.empty(application)){
-                app = application;
+            if (StringUtils.isNotEmpty(controllerRequest.getApplication())) {
+                app = controllerRequest.getApplication();
```
Comment (ofersk) [inline src/main/java/com/checkmarx/flow/controller/ADOController.java:86]: you are calculating this value 2 times. use optional to calculate it a single time

File: src/main/java/com/checkmarx/flow/controller/ADOController.java [hunk 5]
```diff
@@ -50,86 +45,60 @@
...
             }
 
             BugTracker.Type bugType = BugTracker.Type.ADOPULL;
-            if(!ScanUtils.empty(bug)){
-                bugType = ScanUtils.getBugTypeEnum(bug, flowProperties.getBugTrackerImpl());
+            if (StringUtils.isNotEmpty(controllerRequest.getBug())) {
+                bugType = ScanUtils.getBugTypeEnum(controllerRequest.getBug(), flowProperties.getBugTrackerImpl());
```
Comment (ofersk) [inline src/main/java/com/checkmarx/flow/controller/ADOController.java:91]: getBug is calculated 2 times

File: src/main/java/com/checkmarx/flow/controller/ADOController.java [hunk 6]
```diff
@@ -50,86 +45,60 @@
...
             }
 
-            if(appOnlyTracking != null){
-                flowProperties.setTrackApplicationOnly(appOnlyTracking);
+            if (controllerRequest.getAppOnly() != null) {
+                flowProperties.setTrackApplicationOnly(controllerRequest.getAppOnly());
```
Comment (ofersk) [inline src/main/java/com/checkmarx/flow/controller/ADOController.java:95]: getAppOnly is calculated 2 times

File: src/main/java/com/checkmarx/flow/CxFlowRunner.java (hunk 1 above)
Comment (alex-ko-dev) [reply src/main/java/com/checkmarx/flow/CxFlowRunner.java:4]: IntelliJ has a functionality that allows to quickly optimize imports with a single hotkey. It produces * when it considers this relevant. Making manual changes in the import statements will take much time and will be overwritten by the next invocation of 'optimize imports' anyway.

File: src/main/java/com/checkmarx/flow/config/FlowConfig.java (hunk 2 above)
Comment (alex-ko-dev) [reply src/main/java/com/checkmarx/flow/config/FlowConfig.java:15]: Yes, this is to replace Charset.forName("UTF-8")
with StandardCharsets.UTF_8, which looks neater to me.

File: src/main/java/com/checkmarx/flow/controller/ADOController.java (hunk 4 above)
Comment (alex-ko-dev) [reply src/main/java/com/checkmarx/flow/controller/ADOController.java:86]: After a short investigation, using `Optional` vs `if` seems more like a matter of taste to me. Here are some findings:
https://codeburst.io/replace-your-if-else-if-nested-if-not-null-then-get-blocks-with-java-8-optional-s-ed9695a6331e
https://stackoverflow.com/a/56235329
https://codereview.stackexchange.com/a/223329

As for calling getters twice - do you mean the issue of performance? This seems like a micro-optimization to me. One may also argue that the creation of Optional is an expensive operation.

**Pull request merged**